File: backend/ria_ml/market_process.py
import numpy as np

def market_process(market, client):
    nrows, ncols = client['pStatesM'].shape

    u = market['eC']
    v = market['sdC']**2
    b = np.sqrt(np.log(v / u**2 + 1))
    a = 0.5 * np.log(u**2 / np.exp(b**2))
    market['csM'] = np.exp(a + b * np.random.randn(nrows, ncols))
    m = np.cumprod(market['csM'], axis=1)
    market['cumCsM'] = np.hstack([np.ones((nrows, 1)), m[:, :-1]])

    market['rfsM'] = market['rf'] * np.ones((nrows, ncols))
    m = np.cumprod(market['rfsM'], axis=1)
    market['cumRfsM'] = np.hstack([np.ones((nrows, 1)), m[:, :-1]])

    u = market['exRm'] + (market['rf'] - 1)
    v = market['sdRm']**2
    b = np.sqrt(np.log(v / u**2 + 1))
    a = 0.5 * np.log(u**2 / np.exp(b**2))
    market['rmsM'] = np.exp(a + b * np.random.randn(nrows, ncols))
    m = np.cumprod(market['rmsM'], axis=1)
    market['cumRmsM'] = np.hstack([np.ones((nrows, 1)), m[:, :-1]])

    b = np.log(u / market['rf']) / np.log(1 + market['sdRm']**2 / u**2)
    a = np.sqrt(u * market['rf'])**(b - 1)
    as_ = np.ones((nrows, 1)) * (a**np.arange(ncols))
    market['ppcsM'] = as_ * (market['cumRmsM']**(-b))
    market['pvsM'] = market['ppcsM'] / nrows
    market['avec'] = as_[0, :]
    market['b'] = b

    return market

File: backend/ria_ml/market_process.m
function market = market_process(market,client)
  % get size for all matrices from client.pStatesM
     [nrows, ncols] = size(client.pStatesM);
     
  % compute cost of living (inflation) matrix
       u = market.eC; 
       v = market.sdC^2;
       b = sqrt(log((v/(u^2)) + 1)); 
       a = 0.5 *log((u^2)/exp(b^2));
       market.csM = exp(a + b*randn(nrows,ncols));
  % compute cumulative cost of living (inflation) matrix
       m = cumprod(market.csM,2);
       market.cumCsM = [ones(nrows,1) m(:,1:ncols-1)];
       
  % compute risk-free real returns matrix
       market.rfsM = market.rf*ones(nrows,ncols);
  % compute cumulative risk-free real returns matrix at ends of each year
       m = cumprod(market.rfsM,2);
       market.cumRfsM = [ones(nrows,1) m(:,1:ncols-1)];
       
  % compute market returns matrix   
       u = market.exRm + (market.rf - 1); % get total market expected return
       v = market.sdRm^2;
       b = sqrt(log((v/(u^2)) + 1)); 
       a = 0.5 *log((u^2)/exp(b^2));
       market.rmsM = exp(a + b*randn(nrows,ncols));
  % compute market cumulative returns matrix
       m = cumprod(market.rmsM,2);
       market.cumRmsM = [ones(nrows,1) m(:,1:ncols-1)];

  % compute ppcs and present values matrix   
       b = log(u/market.rf)/log(1 + (market.sdRm^2)/(u^2));
       a = sqrt(u*market.rf)^(b-1);
       as = ones(nrows,1)*(a.^(0:ncols-1));
       market.ppcsM = as.*(market.cumRmsM.^-b);
       market.pvsM = market.ppcsM/nrows;
   % temporary
       market.avec = as(1,:);
       market.b = b;
       
 end              